pypesto/objective/roadrunner/roadrunner_calculator.py

```diff
@@ -0,0 +1,438 @@
+        mode:
+            Mode of the call.
+        roadrunner_instance:
+            RoadRunner instance.
+        edatas:
+            List of ExpData.
+        x_ids:
+            Sequence of parameter IDs.
+        parameter_mapping:
+            Parameter parameter_mapping.
+        petab_problem:
+            PEtab problem.
+        solver_options:
+            Solver options of the roadrunner instance Integrator.
+
+        Returns
+        -------
+        Tuple of objective function values.
+        """
+        # sanitiy check that edatas and conditions are consistent
+        if len(edatas) != len(parameter_mapping):
+            raise ValueError(
+                "Number of edatas and conditions are not consistent."
+            )
+        if solver_options is None:
+            solver_options = SolverOptions()
+        # apply solver options
+        solver_options.apply_to_roadrunner(roadrunner_instance)
+        simulation_results = {}
+        llh_tot = 0
+        for edata, mapping_per_condition in zip(edatas, parameter_mapping):
+            sim_res, llh = self.simulate_per_condition(
+                x_dct, roadrunner_instance, edata, mapping_per_condition
+            )
+            simulation_results[edata.condition_id] = sim_res
+            llh_tot += llh
+
+        if mode == MODE_FUN:
+            return {
+                "fval": -llh_tot,
+                "simulation_results": simulation_results,
+                "llh": llh_tot,
+            }
+        if mode == MODE_RES:  # TODO: speed up by not using pandas
+            simulation_df = simulation_to_measurement_df(
+                simulation_results, petab_problem.measurement_df
+            )
+            res_df = petab.calculate_residuals(
+                petab_problem.measurement_df,
+                simulation_df,
+                petab_problem.observable_df,
+                petab_problem.parameter_df,
+            )
+            return {
+                "res": res_df,
+                "simulation_results": simulation_results,
+                "fval": -llh_tot,
+            }
+
+    def simulate_per_condition(
+        self,
+        x_dct: dict,
+        roadrunner_instance: roadrunner.RoadRunner,
+        edata: ExpData,
+        parameter_mapping_per_condition: ParMappingDictQuadruple,
+    ):
+        """Simulate the model for a single condition.
+
+        Parameters
+        ----------
+        x_dct:
+            Parameter dictionary.
+        roadrunner_instance:
+            RoadRunner instance.
+        edata:
+            ExpData of a single condition.
+        parameter_mapping_per_condition:
+            Parameter parameter_mapping for a single condition.
+        """
+        # get timepoints and outputs to simulate
+        timepoints = edata.get_timepoints()
+        if timepoints[0] != 0.0:
+            timepoints = [0.0] + timepoints
+        if len(timepoints) == 1:
+            timepoints = [0.0] + timepoints
```

Comment:
It is because of how roadrunner is constructed. As it has flexible meaning to time, the first value is the one where roadrunner starts its simulation (therefore we need 0 as the first value) and it also means it needs a second value to know up till where it should simulate

Reply:
So `[0.0]` will not suffice to get the initial state?

Reply:
no, [0.0] is not really possible. Should I catch that here? 🤔

Reply:
Ok. No, no change required.